### ISSUE TITLE
dsolve on a vector gives Python exception

### ISSUE THREAD

**Issue by winkmal** [cbm755/octsympy#1040]
Occasionally, I have _octsympy_ run some symbolic scripts (I used to have sporadical access to Matlab Symbolic Toolbox, but not at the moment).
I am quite sure this code used to work, both in Matlab and Octave:

```
try
    % if executed in Octave
     pkg load symbolic
end

%% Define variables for diff. equation
syms x(t) y(t) z(t)
syms x_0 y_0 z_0

%% Solve diff. equation
diffEqs	 = [diff(x, t) == -x + 1, diff(y, t) == -y, diff(z, t) == -z];
initCond = [x(0) == x_0, y(0) == y_0, z(0) == z_0];

solution = dsolve(diffEqs, initCond)
```

However, with Octave 6.1.0, symbolic 2.9.0 and SymPy 1.4, I am getting:

```
error: Python exception: AttributeError: MutableDenseMatrix has no attribute rhs.
    occurred at line 5 of the Python code block:
    ics2[s.lhs] = s.rhs
error: called from
    pycall_sympy__ at line 178 column 7
    dsolve at line 187 column 8
    diffEquation_SymbolicVector at line 15 column 10
```
Is this a regression? Or an error in my code? How can this be fixed?

**Comment by feralber** [comment]
I have tried the code in Matlab and it works. So, it looks like it's a bug in octsympy.

**Comment by rotton** [comment]
Anyone? I have also tried it on Debian 10 with Octave 4.4.1, symbolic 2.7.1 and SymPy 1.3. Same error.

**Comment by mtmiller** [comment]
If you can identify an older version of Octave and/or the symbolic package where this example used to work, that would help identify this as a regression. Otherwise it may be something that has never worked in Octave, and may be a missing feature of either Octave or SymPy.

**Comment by latot** [comment]
Yes, I think I know why this error, here the code from dsolve:

```
  cmd = { 'ode=_ins[0]; ics=_ins[1:]'
          '# convert our input to a dict'
          'ics2 = {}'
          'for s in ics:'
          '    ics2[s.lhs] = s.rhs'
          'sol = sp.dsolve(ode, ics=ics2)'
          'return sol,' };
```

I think the problem is that.., ics travel all our inputs of conditions, but the first condition is an array with the conditions, maybe this can fix it:

```
  cmd = { 'ode=_ins[0]; ics=_ins[1:]'
          '# convert our input to a dict'
          'ics2 = {}'
          'def format(ics2, ics):'
          '    for s in ics:'
          '        if is instance(s, list):'
          '            format(ics2, s)'
          '        else:'
          '            ics2[s.lhs] = s.rhs'
          'format(ics2, ics)
          'sol = sp.dsolve(ode, ics=ics2)'
          'return sol,' };
```

This can be a nice oportunity for a better check of the instances, instead the ```else```.
(well, I don't know if the code works, but I think works to can explain this :3)

Thx.

**Comment by rotton** [comment]
> 
> 
> If you can identify an older version of Octave and/or the symbolic package where this example used to work, that would help identify this as a regression. Otherwise it may be something that has never worked in Octave, and may be a missing feature of either Octave or SymPy.

Well, I am using Octave on my corporate PC, and updates are handled by the company. So I am not able to downgrade it here, but I am sure it used to work about a year ago.
Also, would I need to downgrade Octave and/or Symbolic and/or SymPy? The only clean way I can think of is to install it in an older Linux Version (e.g. [Debian 8 or 9](https://distrowatch.com/table.php?distribution=debian&pkglist=true&version=9#pkglist)) on a virtual machine. Not sure if I will find the time for this anytime soon.

**Comment by rotton** [comment]
> Anyone? I have also tried it on Debian 10 with Octave 4.4.1, symbolic 2.7.1 and SymPy 1.3. Same error.

Just tried it on a Debian 9 live system with Octave 4.0.3, symbolic 2.4.0 and SymPy 1.0
Though looking similar, the error message reads slightly different:

```
error: Python exception: AttributeError: MutableDenseMatrix has no attribute lhs.
    occurred at line 8 of the Python code block
error: called from
    python_cmd at line 176 column 5
    dsolve at line 224 column 10
    testSymbolic at line 14 column 10
```

**Comment by rotton** [comment]
> Yes, I think I know why this error, here the code from dsolve:

Thanks for your comment. Could you please link directly to the specific line of code in the repo?
Also, how would I apply your fix? Just manually swapping the existing code with the one you provided?

**Comment by feralber** [comment]
@latos, I tried your fix (after correcting a couple of typos), but it does not work. Same error as before.

**Comment by latot** [comment]
Yes, probably because there can be multiple types of containers, for example, the mutable matrix, non-mutable matrix, list..., etc, etc, we need to travel it if is a container, in part I mean this for checking the instances.

**Comment by feralber** [comment]
I tried to solve the equation directly from Python (version 3.8.5), using sympy 1.4. I wrote this at the python console:
```
In [2]: from sympy import *
In [3]: x,y,z=symbols('x,y,z',cls=Function)
In [4]: t,x0,y0,z0=symbols('t,x0,y0,z0')
In [5]: eq = (Eq(Derivative(x(t),t), -x(t)+1), Eq(Derivative(y(t),t), -y(t)), Eq(Derivative(z(t),t),-z(t)))                                             
In [6]: ics={x(0):x0,y(0):y0,z(0):z0}
In [7]: sol=dsolve(eq,ics=ics)
```
and I got an error:
```
NotImplementedError: Only homogeneous problems are supported, non-homogenous are not supported currently.
```
Then I tried the same with a more recent sympy (version 1.7.1), and it worked. The solution was:
```
>>> pprint(sol)
⎡                 -t                 -t             -t⎤
⎣x(t) = (x₀ - 1)⋅ℯ   + 1, y(t) = y₀⋅ℯ  , z(t) = z₀⋅ℯ  ⎦
```
It seems that you need at least sympy 1.7. However, octsympy only supports up to sympy 1.4.

So we have another problem here. Even if we corrected dsolve so that it would pass the arguments correctly to python, the solution would not be obtained because we are using sympy 1.4.

**Comment by cbm755** [comment]
Was fixed in #1169.
